addon/services/horizon.js

```diff
@@ -0,0 +1,131 @@
+import Ember from 'ember';
+import config from 'ember-get-config';
+import horizon from 'npm:@horizon/client/dist/horizon';
+
+// The horizon object with configuration from users ENV
+// Note: often the DEV environment will be left blank
+// and rather than just let it
+const hzConfig = config.horizon || {};
+const hz = horizon(hzConfig);
+
+const { RSVP: {Promise}, debug, get, typeOf } = Ember;
+const a = Ember.A;
+
+/**
+ * @class Horizon
+ *
+ * Service methods for interacting with Horizon
+ */
+export default Ember.Service.extend({
+  init() {
+    this._super(...arguments);
+    this._collections = []; // TODO: check whether Horizon does collection caching for you
+    this._subscriptions = [];
+    this._watching = [];
+  },
+  willDestroyElement() {
+    this._watching.forEach(s => s.unsubscribe());
+    this._subscriptions.forEach(s => s.unsubscribe()); // TODO: understand lifecycle better
+    hz.disconnect();
+  },
+
+  currentUser: null,  // set by Horizon Observable
+  status: 'init',     // set by Horizon Observable
+  hasAuthToken: null, // set by Horizon Observable
+  raw: false,         // Horizon default; specifies detail/structure in watched changes
+
+  connect() {
+    return new Promise((resolve, reject) => {
+
+      const status = get(this, 'status');
+      if(status === 'connected') {
+        resolve();
+      } else {
+        hz.connect();
+        hz.onReady(() => {
+          this._statusObservable();
+          this._currentUserObservable();
+          debug('horizon client connected');
+          resolve();
+        });
+        hz.onDisconnected(() => {
+          this._disconnected();
+          reject({code: 'disconnected', config: hz});
```

Comment:
So, logically the only way for a disconnect to happen is to have connected first, which means that we would have already resolved the promise. I understand listening for this and then maybe debug logging some info and/or reconnect logic though.